[extractor/mx3] Add extractor

martinxyz opened the pull request:
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

<!--

Explanation of your *pull request* in arbitrary form goes here. Please **make sure the description explains the purpose and effect** of your *pull request* and is worded well enough to be understood. Provide as much **context and examples** as possible

-->

Add a simple, basic extractor for mx3.ch.

(mx3.ch is a site that hosts music uploaded by bands from or in Switzerland. As a first approximation I'd say it's government funded.)

<details open><summary>Template</summary> <!-- OPEN is intentional -->

<!--

# PLEASE FOLLOW THE GUIDE BELOW

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes `[ ]` relevant to your *pull request* (like [x])
- Use *Preview* tab to see how your *pull request* will actually look like

-->

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

seproDev commented:
I think it would be better to extract metadata from: `https://mx3.ch/t/1LIY.json`
And to also include the other formats:
- `https://mx3.ch/tracks/1Cru/player_asset` (128 kbps mp3)
- `https://mx3.ch/tracks/1Cru/player_asset?quality=hd` (320 kbps mp3)
- `https://mx3.ch/tracks/1Cru/player_asset?quality=` (source wav file)
- `https://mx3.ch/tracks/1C6E/download` (source download. Not available for all files)

martinxyz commented:
Thanks for looking at this. I missed that there is a JSON, using it now where possible. Sadly it is missing the genre info (which I would be tempted to drop, if it was the only thing), and also whether the "download" format is available.

I have added multiple formats now. Previously I've hardcoded the type to "mp3" for non-video, but the Mime-Types are all over the place. (Different video formats, mp3, wav.) I'm not confident about hardcoding a bitrate into the info. I've removed my code that hardcoded the file extension to 'mp3' for audio and 'mp4' for video.

I tried to set 'ext' to None (in the hope to trigger the default code that makes a HEAD request, then guesses the extension) but the testing framework doesn't seem to like that. So I've added the HEAD request directly into the extractor now. It works, but I wonder if there is a better way. (Ideally, the result of the HEAD request would also fill the file size info, etc. - But maybe it's too much, let's first have a working extractor.)

seproDev commented on yt_dlp/extractor/mx3.py:
```diff
@@ -0,0 +1,98 @@
+        webpage = self._download_webpage(url, track_id)
+        json = self._download_json(f'https://mx3.ch/t/{track_id}.json', track_id)
+
+        title = json['title']
+        artist = json.get('artist')
+        if artist and not title.startswith(artist):
+            title = artist + ' - ' + title
+
```
With those changes, this can be removed.
```suggestion
```
I see you are prepending the artists to the title here. I'd say it is probably better to keep the title untouched.
I could potentially see removing the artist name from the title if the site consistently prepends it.
That way, the user can use an output template like this `%(artist)s - %(title)s.%(ext)s`  if they want such filenames.

martinxyz replied:
You're right, leaving the title untouched now. One of the tests has a title starting with the artist's name, but it's a rare exception (and probably intentional).

I'll adapt the template you suggested, thanks! To be honest I never bothered to RTFM until now, but I was always annoyed when the artist was missing from the filename and renamed manually. With those one-word song titles I'd never find anything again.

seproDev commented on yt_dlp/extractor/mx3.py:
```diff
@@ -0,0 +1,98 @@
+                fmt['ext'] = urlhandle_detect_ext(urlh)
+                formats.append(fmt)
+
+        add_format({
+            'url': 'https://mx3.ch/' + json['url'],
+            'format_id': 'default',
+            'quality': 1,
+        })
+
+        if 'hd_url' in json:
+            add_format({
+                'url': 'https://mx3.ch/' + json['hd_url'],
+                'format_id': 'hd',
+                'quality': 10,
+            })
+
+        # the "download" feature is not available everywhere
+        if f'/tracks/{track_id}/download' in webpage:
+            add_format({
+                'url': f'https://mx3.ch/tracks/{track_id}/download',
+                'format_id': 'download',
+                'quality': 11,
+                'format_note': 'usually uncompressed WAV',
+            })
```
I wonder if given that we need to make a HEAD request anyway, we should just attempt to add all formats:
```suggestion
        add_format({
            'url': f'https://mx3.ch/tracks/{track_id}/player_asset',
            'format_id': 'default',
            'quality': 1,
        })

        add_format({
            'url': f'https://mx3.ch/tracks/{track_id}/player_asset?quality=hd',
            'format_id': 'hd',
            'quality': 10,
        })

        add_format({
            'url': f'https://mx3.ch/tracks/{track_id}/download',
            'format_id': 'download',
            'quality': 11,
        })

        add_format({
            'url': f'https://mx3.ch/tracks/{track_id}/player_asset?quality=',
            'format_id': 'download_alternative,
            'quality': 11,
        })
```

martinxyz replied:
Always trying a HEAD seems reasonable to me, done in the latest commits.

I'm not adding the "download_alternative" format because it looks like it would then be picked by default. It seems to be exposed unintentionally, so I don't want to fetch it without confirmation from the user. The files are huge that it borders on abuse, and I'd never want myself. For example https://neo.mx3.ch/t/1hpd would give you a 233MiB WAV by default, while the current choices with 12MiB and 30MiB (hd) are much more reasonable.

seproDev commented on yt_dlp/extractor/mx3.py:
```diff
@@ -0,0 +1,98 @@
+        def add_format(fmt):
+            urlh = self._request_webpage(HEADRequest(fmt['url']), track_id, note='Fetching media headers', fatal=False)
+            if urlh:
+                fmt['ext'] = urlhandle_detect_ext(urlh)
```
You can add the file size:
```suggestion
                fmt['ext'] = urlhandle_detect_ext(urlh)
                fmt['filesize'] = int_or_none(urlh.headers.get('Content-Length'))
```
import `int_or_none` from `..utils`

martinxyz replied:
I'm not too eager to know the filesize specifically, it was just a thought that maybe there is some generic HEAD method that will add all information from that (extension/mime type, file size, date, ...) automatically in one call. But I guess it's not that simple, and not sure if it's a important for users.

seproDev replied:
I would still add the filesize, so the user can see the filesize and bitrate if they look at `-F`

seproDev commented on yt_dlp/extractor/mx3.py:
```diff
@@ -0,0 +1,98 @@
+
+    def _real_extract(self, url):
+        track_id = self._match_id(url)
+        webpage = self._download_webpage(url, track_id)
```
Actually, looking around the page, there is a bunch more optional data on the webpage that isn't in the JSON.
If you like, you can also try to extract those.
`Year of creation` -> `release_year`
`single-stats` -> `view_count`
`Added on` -> `upload_date`
`Description` -> `description` (not always present)
`Tag` -> `tags` (not always present)
Maybe there is more I didn't see.

martinxyz replied:
Who consumes all this information, though? Descriptions and tags sound useful enough to extract, maybe I'll add those, but they won't be embedded into the downloaded MP3s or anything by default, right?

seproDev replied:
Some of this info would be embedded if you use `--embed-metadata`.
Some info may also be useful if you want to filter certain downloads, say you wanted to only download tracks uploaded within the last month.
If you don't think it is worth the hassle to implement, you don't need to do this.

martinxyz replied:
I see; I've added the date now from the HEAD request because it was easy. Yes I think it's diminishing returns here, it looks like we'd have to hunt those down individually in the HTML using nearby translatable strings as an anchor. Seems pretty fragile, too.

seproDev commented on yt_dlp/extractor/mx3.py:
```diff
@@ -0,0 +1,98 @@
+
+
+class Mx3IE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?mx3\.ch/t/(?P<id>[0-9A-Za-z]+)'
```
There also exists https://neo.mx3.ch/ and https://volksmusik.mx3.ch/
I think the extractor will also need to be slightly adjusted to work with both.

martinxyz replied:
Ah I didn't explore those until now - they have quite some additional content. I've added them both.

seproDev replied:
You are now subclassing from a concrete IE, which should not be done.
Instead, I would add another capture group for the subdomain and use that for the api calls.

martinxyz replied:
Capture group sounds like an elegant solution, but this means the 'id' needs to contain the domain too, because it seems like the sites do not share a common id namespace (see my commit messages). That's probably okay? Otherwise I'll just do an abstract base class.

seproDev replied:
Ah right, forgot about that. Yeah, then it's better to move everything to an abstract base class.

martinxyz commented on yt_dlp/extractor/mx3.py:
```diff
@@ -0,0 +1,98 @@
+            'title': title,
+            'artist': artist,
+            'genre': genre,
+            'thumbnail': json.get('picture_url_xlarge') or json.get('picture_url'),
```
With my recent changes the performer_name is now added to the list of artists so this won't work, but it may simplify the code for the other fields. I'll look at this later.

martinxyz commented on yt_dlp/extractor/mx3.py:
```diff
@@ -0,0 +1,160 @@
+                'composer': ('composer_name', {str}),
+                'thumbnail': (('picture_url_xlarge', 'picture_url'), {url_or_none}),
+            }, get_all=False),
+        }
```
The artist/album_artist split seem to fit pretty well, I like it. I slightly preferred the filenames I got previously with the format string, I'll get a few duplicated artist names now but it's not too bad really.

I've updated the tests to match. I noticed that https://neo.mx3.ch/t/1hpd kind of has a description, but they put it all into the credits field, not sure if we want to add that.

martinxyz commented on yt_dlp/extractor/mx3.py:
```diff
@@ -0,0 +1,160 @@
+                urlh = self._request_webpage(HEADRequest(fmt['url']), track_id, note='Fetching default media headers')
+            else:
+                urlh = self._request_webpage(HEADRequest(fmt['url']), track_id, fatal=False, expected_status=404,
+                                             note=f'Trying media headers for optional format {fmt["format_id"]}')
```
My idea was that you'd then see the HTTP error (if there is one), but... Done. I like the cleaner code too, and you'd have to investigate anyway if it goes wrong.

martinxyz commented on yt_dlp/extractor/mx3.py:
```diff
@@ -0,0 +1,160 @@
+            'url': f'{track_url}/download',
+            'format_id': 'download',
+            'quality': 11,
+        }, fatal=False)
```
Hm. I can see your point. But for me there is simply no difference in quality between a high-bitrate MP3 and WAV. For video I would list the formats and pick one manually, but for music I use the "Open With" browser extension (non-interactive) and check the download folder later. I'm going to use `-f"best[ext!=wav][ext!=flac][filesize<50M]" -x` now, so it will work for me if you add the format.

seproDev replied:
If you only want the format selection for this site, just use `-f hd/default`. This will download `hd` if available and otherwise fallback to `default`.

bashonly commented:
Are there overlapping IDs between the 3 sites? Could this just be 1 extractor?

seproDev commented:
There are collisions [1g2T (neo)](https://neo.mx3.ch/t/1g2T),  [1g2T (volksmusik)](https://volksmusik.mx3.ch/t/1g2T)
